src/waveform/renderers/allshader/waveformrenderertextured.cpp

```diff
@@ -0,0 +1,515 @@
+
+#include <QOpenGLFramebufferObject>
+#include <QOpenGLShaderProgram>
+
+#include "moc_waveformrenderertextured.cpp"
+#include "track/track.h"
+#include "waveform/renderers/waveformwidgetrenderer.h"
+#include "waveform/waveform.h"
+
+namespace allshader {
+
+// static
+QString WaveformRendererTextured::fragShaderForType(WaveformRendererTextured::Type t) {
+    using Type = WaveformRendererTextured::Type;
+    switch (t) {
+    case Type::Filtered:
+        return QLatin1String(":/shaders/filteredsignal.frag");
+    case Type::RGB:
+        return QLatin1String(":/shaders/rgbsignal.frag");
+    case Type::Stacked:
+        return QLatin1String(":/shaders/stackedsignal.frag");
+    default:
+        break;
+    }
+    assert(false);
+    return QString();
+}
+
+WaveformRendererTextured::WaveformRendererTextured(WaveformWidgetRenderer* waveformWidget,
+        Type t)
+        : WaveformRendererSignalBase(waveformWidget),
+          m_unitQuadListId(-1),
+          m_textureId(0),
+          m_textureRenderedWaveformCompletion(0),
+          m_shadersValid(false),
+          m_type(t),
+          m_pFragShader(fragShaderForType(t)) {
+}
+
+WaveformRendererTextured::~WaveformRendererTextured() {
+    if (m_textureId) {
+        glDeleteTextures(1, &m_textureId);
+    }
+
+    if (m_frameShaderProgram) {
+        m_frameShaderProgram->removeAllShaders();
+    }
+}
+
+bool WaveformRendererTextured::loadShaders() {
+    qDebug() << "WaveformRendererTextured::loadShaders";
+    m_shadersValid = false;
+
+    if (m_frameShaderProgram->isLinked()) {
+        m_frameShaderProgram->release();
+    }
+
+    m_frameShaderProgram->removeAllShaders();
+
+    if (!m_frameShaderProgram->addShaderFromSourceFile(
+                QOpenGLShader::Vertex,
+                ":/shaders/passthrough.vert")) {
+        qDebug() << "WaveformRendererTextured::loadShaders - "
+                 << m_frameShaderProgram->log();
+        return false;
+    }
+
+    if (!m_frameShaderProgram->addShaderFromSourceFile(
+                QOpenGLShader::Fragment,
+                m_pFragShader)) {
+        qDebug() << "WaveformRendererTextured::loadShaders - "
+                 << m_frameShaderProgram->log();
+        return false;
+    }
+
+    if (!m_frameShaderProgram->link()) {
+        qDebug() << "WaveformRendererTextured::loadShaders - "
+                 << m_frameShaderProgram->log();
+        return false;
+    }
+
+    if (!m_frameShaderProgram->bind()) {
+        qDebug() << "WaveformRendererTextured::loadShaders - shaders binding failed";
+        return false;
+    }
+
+    m_shadersValid = true;
+    return true;
+}
+
+bool WaveformRendererTextured::loadTexture() {
+    int dataSize = 0;
+    const WaveformData* data = nullptr;
+
+    ConstWaveformPointer pWaveform = m_waveformRenderer->getWaveform();
+    if (pWaveform) {
+        dataSize = pWaveform->getDataSize();
+        if (dataSize > 1) {
+            data = pWaveform->data();
+        }
+    }
```

Comment:
declare variables at point of use, prefer early returns when it doesn't make sense to continue (this is not the linux kernel ;) )
```suggestion
    ConstWaveformPointer pWaveform = m_waveformRenderer->getWaveform();
    if (!pWaveform) {
        return false;
    }
    int dataSize = pWaveform->getDataSize();
    // is this condition correct? should it actually be just be dataSize == 0?
    if (dataSize <= 1) {
        return false;
    }
    const WaveformData* data = pWaveform->data();
```

Reply:
This is not my code.